csharp/ql/src/semmle/code/csharp/AnnotatedType.qll

```diff
@@ -117,22 +125,98 @@ private module Annotations {
 
     /** Gets a textual representation of this type annotation. */
     string toString() { result = getTypePrefix() + getTypeSuffix() }
+
+    private predicate isSet(int bit) {
+      isBit(bit) and
+      exists(int mask | mask = getBitMask(bit) | flags.bitAnd(mask) = mask)
+    }
+
+    /** Gets an annotation in this set of annotations. */
+    TypeAnnotation getAnAnnotation() {
+      isSet(result.getBit())
+      or
+      nullability instanceof AnnotatedNullability and result instanceof NullableRefType
+      or
+      nullability instanceof NotAnnotatedNullability and result instanceof NonNullableRefType
+    }
+  }
+
+  /**
+   * A structured type annotation representing type nullability.
+   * For example, `IDictionary<string!,string?>?` has nullability `<!,?>?`.
+   */
+  class Nullability extends @nullability {
+    string toString() { result = getMemberString() + getSelfNullability() }
+
+    language[monotonicAggregates]
+    private string getMemberString() {
+      if nullability_member(this, _, _)
+      then
+        result = "<" +
+            concat(int i, Nullability child |
+              nullability_member(this, i, child)
+            |
+              child.toString(), "," order by i
+            ) + ">"
+      else result = ""
+    }
+
+    /**
+     * Gets the `i`th member of this annotation.
+     * Returns `this` if the nullability is not explicitly
+     * stored in the database, since many type annotations will have consistent
+     * nullability.
+     */
+    bindingset[i]
+    Nullability getMember(int i) {
```

Comment:
I don't much like exposing this predicate, as it doesn't bind `i`. Perhaps it should just be a private top-level predicate instead?